### PR TITLE
`GoogleDriveHook`: Add folder_id param to upload_file

### LINKED ISSUE #29552: Google provider doesn't let uploading file to a shared drive
### Apache Airflow Provider(s)

google

### Versions of Apache Airflow Providers

```
apache-airflow-providers-google: version 8.9.0
```

### Apache Airflow version

2.5.1

### Operating System

Linux -  official airflow image from docker hub apache/airflow:slim-2.5.1

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### What happened

Not sure if it's a bug or a feature.

Originally I have used `LocalFilesystemToGoogleDriveOperator` to try uploading a file into a shared Google Drive without success.
Provider didn't find a directory with a given name so it created a new one without browsing `shared drives`. Method call that doesn't allow to upload it is here:
https://github.com/apache/airflow/blob/main/airflow/providers/google/suite/hooks/drive.py#L223

### What you think should happen instead

It would be nice if there was an optional parameter to provide a `drive_id` into which user would like to upload a file. With the same directory check behaviour that already exists but extended to `shared drives`

### How to reproduce

1. Create a directory on the shared google drive
2. Fill `LocalFilesystemToGoogleDriveOperator` constructor with the arguments.
3. Execute the function

### Anything else

I am willing to submit a PR but I would need to know more details, your thoughts, expectations of the implementation to make as little iteration on it possible.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by shahar1**
---
replaces: #28142
closes: https://github.com/apache/airflow/issues/29552
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by eladkal — APPROVED** [review]

File: airflow/providers/google/suite/hooks/drive.py [hunk 1]
```diff
@@ -215,14 +226,15 @@ def upload_file(
             or to -1.
         :param resumable: True if this is a resumable upload. False means upload
             in a single request.
+        :param folder_id: The base/root folder id for remote_location (part of the drive url of a folder).
```
Comment (jbampton) [inline airflow/providers/google/suite/hooks/drive.py]:
```suggestion
        :param folder_id: The base/root folder id for remote_location (part of the drive URL of a folder).
```

**Comment by eladkal** [comment]
@shahar1 can you check https://github.com/apache/airflow/issues/29552 I think it would be good to add the parameter to `LocalFilesystemToGoogleDriveOperator`

**Comment by aru-trackunit** [comment]
@shahar1 I pulled your code and only added `folder_id` assignments
```
    file_upload = LocalFilesystemToGoogleDriveOperator(
        task_id="uploading-file"
        gcp_conn_id="gc-conn-id",
        drive_folder="Folder A",
        folder_id="Folder_A_ID",
        local_paths=[f"/tmp/file_to_upload.csv"]
    )
```
The good news is that file was uploaded to a shared drive, however path's are not matching what I meant.
I wanted the file to be uploaded to `/Folder A/file.csv` 
`/Folder_A_ID/Folder A/file.csv`

That was the case when name of the `folder_id` was equal to `drive_folder`. But maybe I expected that because I knew what the behaviour was previously.

In order to get what I wanted I had to run this:
```
    file_upload = LocalFilesystemToGoogleDriveOperator(
        task_id="uploading-file"
        gcp_conn_id="gc-conn-id",
        drive_folder="/",
        folder_id="Folder_A_ID",
        local_paths=[f"/tmp/file_to_upload.csv"]
    )
```

so it solves the problem, but then logs message is quite misleading.

```
[2023-02-15, 11:25:52 UTC] {drive.py:238} INFO - File /tmp/file_to_upload.csv uploaded to gdrive:///file_to_upload.csv.
```
Would be nice to get `folder_id` real path. - Not sure if possible

In case this goes forward the documentation would have to be improved & maybe rename `folder_id` to `parent_folder_id` in this case and `drive_folder` to `path`. But if someone doesn't fill `folder_id` then it gets cumbersome.

When I run:
```
    file_upload = LocalFilesystemToGoogleDriveOperator(
        task_id="uploading-file"
        gcp_conn_id="gc-conn-id",
        drive_folder="Folder A",
        folder_id="root", #I could skip that param, but added to be clear.
        local_paths=[f"/tmp/test-file-csv.csv"]
    )
```
then it's placed on the user's shared drive under `Folder A` - it's a correct behaviour

@eladkal 
But yeah! Overall it's almost there!

**Comment by shahar1** [comment]
> This comment is in regards to this issue: #29552
> 
> @shahar1 I pulled your code and only added `folder_id` assignments
> 
> ```
>     file_upload = LocalFilesystemToGoogleDriveOperator(
>         task_id="uploading-file"
>         gcp_conn_id="gc-conn-id",
>         drive_folder="Folder A",
>         folder_id="Folder_A_ID",
>         local_paths=[f"/tmp/file_to_upload.csv"]
>     )
> ```
> 
> The good news is that file was uploaded to a shared drive, however path's are not matching what I meant. I wanted the file to be uploaded to `/Folder A/file.csv`, but it resulted in `/Folder_A_ID/Folder A/file.csv`
> 
> That was the case when name of the `folder_id` was equal to `drive_folder`. But maybe I expected that because I knew what the behaviour was previously.
> 
> In order to get what I wanted I had to run this:
> 
> ```
>     file_upload = LocalFilesystemToGoogleDriveOperator(
>         task_id="uploading-file"
>         gcp_conn_id="gc-conn-id",
>         drive_folder="/",
>         folder_id="Folder_A_ID",
>         local_paths=[f"/tmp/file_to_upload.csv"]
>     )
> ```
> 
> so it solves the problem, but then logs message is quite misleading.
> 
> ```
> [2023-02-15, 11:25:52 UTC] {drive.py:238} INFO - File /tmp/file_to_upload.csv uploaded to gdrive:///file_to_upload.csv.
> ```
> 
> Would be nice to get `folder_id` real path. - Not sure if possible
> 
> In case this goes forward the documentation would have to be improved & maybe rename `folder_id` to `parent_folder_id` in this case and `drive_folder` to `path`. But if someone doesn't fill `folder_id` then it gets cumbersome.
> 
> When I run:
> 
> ```
>     file_upload = LocalFilesystemToGoogleDriveOperator(
>         task_id="uploading-file"
>         gcp_conn_id="gc-conn-id",
>         drive_folder="Folder A",
>         folder_id="root", #I could skip that param, but added to be clear.
>         local_paths=[f"/tmp/test-file-csv.csv"]
>     )
> ```
> 
> then it's placed on the user's own drive under `Folder A` - it's a correct behaviour
> 
> @eladkal But yeah! Overall it's almost there!

Hey, thanks for the suggestion - I just added the `folder_id` as a parameter of the operator with a default `root` value.

**Review by eladkal — APPROVED** [review]

**Pull request merged**